### PR TITLE
ISSUE-496 Implement FirebaseSubscription/set

### PR DESCRIPTION
**Pull request by vttranlina**
resolve https://github.com/linagora/tmail-backend/issues/496

### CODE REVIEW

File: tmail-backend/integration-tests/jmap/jmap-integration-tests-common/src/main/scala/com/linagora/tmail/james/common/FirebaseSubscriptionSetMethodContract.scala [hunk 1]
```diff
@@ -0,0 +1,868 @@
...
+      .addUser(ANDRE.asString(), ANDRE_PASSWORD)
+
+    requestSpecification = baseRequestSpecBuilder(server)
+      .setAuth(authScheme(UserCredential(BOB, BOB_PASSWORD)))
+      .addHeader(ACCEPT.toString, ACCEPT_RFC8621_VERSION_HEADER)
+      .build()
+  }
+
+  @Test
+  def setShouldReturnCreatedWhenInvalidCreationRequest(): Unit = {
```
Comment (Arsnael) [inline tmail-backend/integration-tests/jmap/jmap-integration-tests-common/src/main/scala/com/linagora/tmail/james/common/FirebaseSubscriptionSetMethodContract.scala:38]:
```suggestion
  def setShouldReturnCreatedWhenValidCreationRequest(): Unit = {
```

File: tmail-backend/integration-tests/jmap/jmap-integration-tests-common/src/main/scala/com/linagora/tmail/james/common/FirebaseSubscriptionSetMethodContract.scala [hunk 2]
```diff
@@ -0,0 +1,868 @@
...
+      s"""{
+         |    "sessionState": "${SESSION_STATE.value}",
+         |    "methodResponses": [
+         |        [
+         |            "FirebaseRegistration/set",
+         |            {
+         |                "created": {
+         |                    "4f29": {
+         |                        "id": "$${json-unit.ignore}",
+         |                        "expires":  "${expires}"
```
Comment (Arsnael) [inline tmail-backend/integration-tests/jmap/jmap-integration-tests-common/src/main/scala/com/linagora/tmail/james/common/FirebaseSubscriptionSetMethodContract.scala:463]: Hm I'm pretty sure usually with jmap method specs if you provide a property in the request and that one does not change (like here for example, as the expiration is not above the max), you should not return the property "expires" here.

If it is above the max though, then the server sets it to the max -> different value than in request -> returned in response (need test for that case as well?)

File: tmail-backend/jmap/extensions/src/main/scala/com/linagora/tmail/james/jmap/json/FirebaseSubscriptionSerializer.scala [hunk 3]
```diff
@@ -35,15 +39,53 @@ object FirebaseSubscriptionSerializer {
...
+  private implicit val mapCreationRequestBySubscriptionCreationId: Reads[Map[FirebaseSubscriptionCreationId, JsObject]] =
+    Reads.mapReads[FirebaseSubscriptionCreationId, JsObject] { string =>
+      refineV[IdConstraint](string)
+        .fold(e => JsError(s"firebase subscription id needs to match id constraints: $e"),
+          id => JsSuccess(FirebaseSubscriptionCreationId(id)))
+    }
+
+  private implicit val patchObject: Reads[FirebaseSubscriptionPatchObject] = Json.valueReads[FirebaseSubscriptionPatchObject]
+
+  private implicit val mapUpdateRequestBySubscriptionCreationId: Reads[Map[UnparsedFirebaseSubscriptionId, FirebaseSubscriptionPatchObject]] =
```
Comment (Arsnael) [inline tmail-backend/jmap/extensions/src/main/scala/com/linagora/tmail/james/jmap/json/FirebaseSubscriptionSerializer.scala:51]: I don't think this work is about set update yet :)

File: tmail-backend/jmap/extensions/src/main/scala/com/linagora/tmail/james/jmap/json/FirebaseSubscriptionSerializer.scala [hunk 4]
```diff
@@ -35,15 +39,53 @@ object FirebaseSubscriptionSerializer {
...
         "types" -> response.types)
 
+  private implicit val mapCreationRequestBySubscriptionCreationId: Reads[Map[FirebaseSubscriptionCreationId, JsObject]] =
+    Reads.mapReads[FirebaseSubscriptionCreationId, JsObject] { string =>
+      refineV[IdConstraint](string)
+        .fold(e => JsError(s"firebase subscription id needs to match id constraints: $e"),
+          id => JsSuccess(FirebaseSubscriptionCreationId(id)))
+    }
+
+  private implicit val patchObject: Reads[FirebaseSubscriptionPatchObject] = Json.valueReads[FirebaseSubscriptionPatchObject]
```
Comment (Arsnael) [inline tmail-backend/jmap/extensions/src/main/scala/com/linagora/tmail/james/jmap/json/FirebaseSubscriptionSerializer.scala:49]: Idem

File: tmail-backend/jmap/extensions/src/main/scala/com/linagora/tmail/james/jmap/model/FirebaseSubscription.scala [hunk 5]
```diff
@@ -107,3 +123,56 @@ object FirebaseSubscriptionGetResponse {
 
 case class FirebaseSubscriptionGetResponse(list: Seq[FirebaseSubscription],
                                            notFound: Set[UnparsedFirebaseSubscriptionId] = Set())
+
+case class FirebaseSubscriptionCreationId(id: Id) {
+  def serialise: String = id.value
+}
+
+case class FirebaseSubscriptionPatchObject(value: Map[String, JsValue])
```
Comment (Arsnael) [inline tmail-backend/jmap/extensions/src/main/scala/com/linagora/tmail/james/jmap/model/FirebaseSubscription.scala:131]: Idem

File: tmail-backend/jmap/extensions/src/main/scala/com/linagora/tmail/james/jmap/model/FirebaseSubscription.scala [hunk 6]
```diff
@@ -107,3 +123,56 @@ object FirebaseSubscriptionGetResponse {
...
+
+case class FirebaseSubscriptionSetRequest(create: Option[Map[FirebaseSubscriptionCreationId, JsObject]],
+                                          update: Option[Map[UnparsedFirebaseSubscriptionId, FirebaseSubscriptionPatchObject]],
+                                          destroy: Option[Seq[UnparsedFirebaseSubscriptionId]]) extends WithoutAccountId
```
Comment (Arsnael) [inline tmail-backend/jmap/extensions/src/main/scala/com/linagora/tmail/james/jmap/model/FirebaseSubscription.scala:135]: create is enough for now

File: tmail-backend/jmap/extensions/src/main/scala/com/linagora/tmail/james/jmap/json/FirebaseSubscriptionSerializer.scala (hunk 3 above)
Comment (quantranhong1999) [reply tmail-backend/jmap/extensions/src/main/scala/com/linagora/tmail/james/jmap/json/FirebaseSubscriptionSerializer.scala:51]: Or we can let Tung crashes the sprint :p

File: tmail-backend/jmap/extensions/src/main/scala/com/linagora/tmail/james/jmap/method/FirebaseSubscriptionSetCreatePerformer.scala [hunk 7]
```diff
@@ -0,0 +1,82 @@
...
+
+  private def parseCreate(json: JsObject): Either[Exception, FirebaseSubscriptionCreationRequest] = for {
+    validJsObject <- FirebaseSubscriptionCreation.validateProperties(json)
+    parsedRequest <- serializer.deserializeFirebaseSubscriptionCreationRequest(validJsObject).asEither
+      .left.map(errors => FirebaseSubscriptionCreationParseException.from(errors))
+    validatedRequest <- parsedRequest.validate
+      .left.map(e => FirebaseSubscriptionCreationParseException(SetError.invalidArguments(SetErrorDescription(e.getMessage))))
+  } yield {
+    validatedRequest
+  }
```
Comment (quantranhong1999) [inline tmail-backend/jmap/extensions/src/main/scala/com/linagora/tmail/james/jmap/method/FirebaseSubscriptionSetCreatePerformer.scala:71]: indent

File: tmail-backend/jmap/extensions/src/main/scala/com/linagora/tmail/james/jmap/method/FirebaseSubscriptionSetCreatePerformer.scala [hunk 8]
```diff
@@ -0,0 +1,82 @@
...
+
+  private def create(clientId: FirebaseSubscriptionCreationId, request: FirebaseSubscriptionCreationRequest, username: Username): SMono[CreationResult] =
+    SMono.fromPublisher(repository.save(username, request))
+      .map(subscription => CreationSuccess(clientId, FirebaseSubscriptionCreationResponse(subscription.id, showExpires(subscription.expires, request))))
+      .onErrorResume(e => SMono.just[CreationResult](CreationFailure(clientId, e)))
```
Comment (quantranhong1999) [inline tmail-backend/jmap/extensions/src/main/scala/com/linagora/tmail/james/jmap/method/FirebaseSubscriptionSetCreatePerformer.scala:76]: I would love to see us validate the input token here (only available in https://github.com/linagora/tmail-backend/pull/519 though)

Comment (vttranlina) [reply tmail-backend/jmap/extensions/src/main/scala/com/linagora/tmail/james/jmap/method/FirebaseSubscriptionSetCreatePerformer.scala:76]: I just looked in the firebase library. It hasn't token validation method (client side). 
Should we post a request to FCM to validate the token?

Comment (quantranhong1999) [reply tmail-backend/jmap/extensions/src/main/scala/com/linagora/tmail/james/jmap/method/FirebaseSubscriptionSetCreatePerformer.scala:76]: We have: https://github.com/linagora/tmail-backend/pull/519#discussion_r1014977239
However, we are discussing whether to use it.

File: tmail-backend/jmap/extensions/src/main/java/com/linagora/tmail/james/jmap/firebase/FirebaseCommonModule.java [hunk 9]
```diff
@@ -2,12 +2,12 @@
...
 import com.linagora.tmail.james.jmap.method.FirebaseCapabilitiesModule;
-import com.linagora.tmail.james.jmap.method.FirebaseSubscriptionGetMethodModule;
+import com.linagora.tmail.james.jmap.method.FirebaseSubscriptionMethodModule;
 
 public class FirebaseCommonModule extends AbstractModule {
     @Override
     protected void configure() {
         install(new FirebaseCapabilitiesModule());
-        install(new FirebaseSubscriptionGetMethodModule());
+        install(new FirebaseSubscriptionMethodModule());
```
Comment (chibenwa) [inline tmail-backend/jmap/extensions/src/main/java/com/linagora/tmail/james/jmap/firebase/FirebaseCommonModule.java:11]: IMO you should keep both ^^

File: tmail-backend/jmap/extensions/src/main/scala/com/linagora/tmail/james/jmap/json/FirebaseSubscriptionSerializer.scala [hunk 10]
```diff
@@ -35,15 +39,44 @@ object FirebaseSubscriptionSerializer {
...
+  private implicit val firebaseSubscriptionSetRequestReads: Reads[FirebaseSubscriptionSetRequest] = Json.reads[FirebaseSubscriptionSetRequest]
+  private implicit val firebaseSubscriptionCreationResponseWrites: Writes[FirebaseSubscriptionCreationResponse] = Json.writes[FirebaseSubscriptionCreationResponse]
+  private implicit val subscriptionMapSetErrorForCreationWrites: Writes[Map[FirebaseSubscriptionCreationId, SetError]] =
+    mapWrites[FirebaseSubscriptionCreationId, SetError](_.serialise, setErrorWrites)
+
+  private implicit val subscriptionMapCreationResponseWrites: Writes[Map[FirebaseSubscriptionCreationId, FirebaseSubscriptionCreationResponse]] =
+    mapWrites[FirebaseSubscriptionCreationId, FirebaseSubscriptionCreationResponse](_.serialise, firebaseSubscriptionCreationResponseWrites)
+
+  private implicit val firebaseSubscriptionSetResponseWrites: OWrites[FirebaseSubscriptionSetResponse] = Json.writes[FirebaseSubscriptionSetResponse]
+  private implicit val subscriptionExpiredTimeReads: Reads[FirebaseSubscriptionExpiredTime] = Json.valueReads[FirebaseSubscriptionExpiredTime]
```
Comment (chibenwa) [inline tmail-backend/jmap/extensions/src/main/scala/com/linagora/tmail/james/jmap/json/FirebaseSubscriptionSerializer.scala:65]: Use `valueFormat instead, to combine valueReads and valueWrites in one line.

Please apply this comment where relevant.

File: tmail-backend/jmap/extensions/src/main/scala/com/linagora/tmail/james/jmap/method/FirebaseSubscriptionSetCreatePerformer.scala [hunk 11]
```diff
@@ -0,0 +1,80 @@
...
+object FirebaseSubscriptionSetCreatePerformer {
+  trait CreationResult
+
+  case class CreationSuccess(clientId: FirebaseSubscriptionCreationId, response: FirebaseSubscriptionCreationResponse) extends CreationResult
+
+  case class CreationFailure(clientId: FirebaseSubscriptionCreationId, e: Throwable) extends CreationResult {
+    def asMessageSetError: SetError = e match {
+      case e: FirebaseSubscriptionCreationParseException => e.setError
+      case e: ExpireTimeInvalidException => SetError.invalidArguments(SetErrorDescription(e.getMessage), Some(Properties("expires")))
+      case e: DeviceClientIdInvalidException => SetError.invalidArguments(SetErrorDescription(e.getMessage))
```
Comment (chibenwa) [inline tmail-backend/jmap/extensions/src/main/scala/com/linagora/tmail/james/jmap/method/FirebaseSubscriptionSetCreatePerformer.scala:26]:
```suggestion
      case e: DeviceClientIdInvalidException => SetError.invalidArguments(SetErrorDescription(e.getMessage), ome(Properties("deviceId"))
```

??

File: tmail-backend/jmap/extensions/src/main/scala/com/linagora/tmail/james/jmap/method/FirebaseSubscriptionSetCreatePerformer.scala [hunk 12]
```diff
@@ -0,0 +1,80 @@
...
+      case e: IllegalArgumentException => SetError.invalidArguments(SetErrorDescription(e.getMessage))
+      case _ => SetError.serverFail(SetErrorDescription(e.getMessage))
```
Comment (chibenwa) [inline tmail-backend/jmap/extensions/src/main/scala/com/linagora/tmail/james/jmap/method/FirebaseSubscriptionSetCreatePerformer.scala:28]: Maybe we need at least a warning log on why this fails...

That's quote generic with all JMAP methods I know but if we have a serverFail we need to have a log about it IMO...

File: tmail-backend/jmap/extensions/src/main/scala/com/linagora/tmail/james/jmap/method/FirebaseSubscriptionSetCreatePerformer.scala (hunk 8 above)
Comment (chibenwa) [reply tmail-backend/jmap/extensions/src/main/scala/com/linagora/tmail/james/jmap/method/FirebaseSubscriptionSetCreatePerformer.scala:74]: > I would love to see us validate the input token here (only available in https://github.com/linagora/tmail-backend/pull/519 though)

Nice remark!

See https://github.com/linagora/tmail-backend/issues/520

**Review by chibenwa — COMMENTED** [review]
Looks good to me.

**Review by Arsnael — APPROVED** [review]

File: tmail-backend/jmap/extensions/src/main/scala/com/linagora/tmail/james/jmap/method/FirebaseSubscriptionSetCreatePerformer.scala [hunk 13]
```diff
@@ -0,0 +1,86 @@
...
+
+  case class CreationSuccess(clientId: FirebaseSubscriptionCreationId, response: FirebaseSubscriptionCreationResponse) extends CreationResult
+
+  case class CreationFailure(clientId: FirebaseSubscriptionCreationId, e: Throwable) extends CreationResult {
+    def asMessageSetError: SetError = e match {
+      case e: FirebaseSubscriptionCreationParseException => e.setError
+      case e: ExpireTimeInvalidException => SetError.invalidArguments(SetErrorDescription(e.getMessage), Some(Properties("expires")))
+      case e: DeviceClientIdInvalidException => SetError.invalidArguments(SetErrorDescription(e.getMessage), Some(Properties("deviceClientId")))
+      case e: IllegalArgumentException => SetError.invalidArguments(SetErrorDescription(e.getMessage))
+      case _ => {
```
Comment (quantranhong1999) [inline tmail-backend/jmap/extensions/src/main/scala/com/linagora/tmail/james/jmap/method/FirebaseSubscriptionSetCreatePerformer.scala:31]: Please handle also `TokenInvalidException` (duplicate FCM token) and a test for it.

**Review by quantranhong1999 — APPROVED** [review]
Otherwise good to me

**Pull request merged**